# multiOmics integration

# useful links
## Tools/Apps/DBs...
* [pISA-tree](https://github.com/NIB-SI/pISA-tree)
* [DiNAR](https://github.com/NIB-SI/DiNAR)
* [Cytoscape](https://cytoscape.org/)
* [R Markdown](https://bookdown.org/yihui/rmarkdown/)
* [mixOmics](https://mixomics.org/)
* [PLAZA](https://ftp.psb.ugent.be/pub/plaza/plaza_public_dicots_05/)
* [MapMan](https://mapman.gabipd.org/)
* [GSEA](https://www.gsea-msigdb.org/gsea/index.jsp)
* [splinecluster](https://www.ma.imperial.ac.uk/~naheard/software/splinecluster/index.html)
* [igraph](https://igraph.org/)
* [SKM](https://skm.nib.si/)
## Articles
* [Misunderstood parameter of NCBI BLAST impacts the correctness of bioinformatics workflows](https://academic.oup.com/bioinformatics/article/35/9/1613/5106166)
* [Methods for the integration of multi-omics data: mathematical aspects](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-015-0857-9)
* [Network approaches to systems biology analysis of complex disease: integrative methods for multi-omics data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6454489/)
* [Canonical Correlation Analysis and Partial Least Squares for Identifying Brain–Behavior Associations: A Tutorial and a Comparative Study](https://www.sciencedirect.com/science/article/pii/S2451902222001859)
* [Visualising associations between paired ‘omics’ data sets](https://biodatamining.biomedcentral.com/articles/10.1186/1756-0381-5-19)
* [mixOmics: An R package for ‘omics feature selection and multiple data integration](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005752)
* [DIABLO: an integrative approach for identifying key molecular drivers from multi-omics assays](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6735831/)
* [Integrating omics datasets with the OmicsPLS package](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2371-3)
## Vignetes/F-ctions/CheatSheets
* [multi-block sPLS-DA](https://mixomicsteam.github.io/mixOmics-Vignette/)
* [RBH search](https://drostlab.github.io/orthologr/reference/blast_rec.html)
* [Biokit](https://github.com/martingarridorc/biokit)
* [clusterProfiler](https://hbctraining.github.io/DGE_workshop_salmon_online/lessons/10_FA_over-representation_analysis.html)
* [fgsea/clusterProfiler](https://pnnl-comp-mass-spec.github.io/proteomics-data-analysis-tutorial/gsea.html)
* [No. clu.](https://stackoverflow.com/questions/15376075/cluster-analysis-in-r-determine-the-optimal-number-of-clusters)
* [caret](https://www.machinelearningplus.com/machine-learning/caret-package/)
* [FS - RF](https://topepo.github.io/caret/recursive-feature-elimination.html)
* [ComplexHeatmap](https://jokergoo.github.io/ComplexHeatmap-reference/book/)
* [rstatix](https://rpkgs.datanovia.com/rstatix/)
* [ggpubr](https://rpkgs.datanovia.com/ggpubr/index.html)
## Lectures
