---
title: "01_ortho"
author: "Maria, zagor"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    self_contained: yes
    fig_width: 12
    fig_height: 9
    toc: yes
    toc_depth: 2
    toc_float: true
    number_sections: yes
    theme: flatly
    highlight: tango
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(#dev = c('pdf', 'png'),  # this embeds pdf and crates scrolable blocks
                      dev = c('png'), 
                      fig.align = 'center', 
                      fig.height = 9, 
                      fig.width = 12 ,
                      warning = FALSE, message = FALSE
                      )
# options(knitr.table.format = "html")

```


1. Download ```BLAST``` for Windows (```.exe``` file) from <https://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/> and install locally

2. Add path to to NCBI blast /bin (could be ```C:\Program Files\NCBI\blast-2.14.0+\bin```) in System Enviroment Variables, e.g. on how to here <https://www.hanss.info/sebastian/post/rtools-path/>

3.Install dependencies ```iocManager::install(c("Biostrings", "GenomicFeatures", "GenomicRanges", "Rsamtools", "IRanges", "rtracklayer", "biomaRt"))```

3. Install ```metablastr``` and ```orthologr``` packages in R

```
devtools::install_github("HajkD/metablastr")

devtools::install_github("HajkD/orthologr")
```


4. Run ```system("blastp -help")``` in R Console to see if it is working



# Libraries

```{r,  echo=TRUE, warning=FALSE, message=FALSE}

rm(list = ls(all = TRUE))
gc()

set.seed(123456)


`%nin%` = Negate(`%in%`)

library(RColorBrewer)
library(magrittr)


# Install package dependencies
# BiocManager::install(c("Biostrings", "GenomicRanges", "GenomicFeatures", "Rsamtools", "rtracklayer"))

# devtools::install_github("HajkD/metablastr")
# devtools::install_github("HajkD/orthologr")

## https://heronoh.github.io/BLASTr
# remotes::install_github("heronoh/BLASTr")
# library(BLASTr)
# devtools::install_github("mhahsler/rBLAST")
# library(rBLAST)

# BiocManager::install(c("Biostrings", "GenomicFeatures", "GenomicRanges", "Rsamtools", "IRanges", "rtracklayer", "biomaRt"))

# install.packages("devtools")
# install the current version of metablastr on your system
# devtools::install_github("HajkD/metablastr", build_vignettes = TRUE, dependencies = TRUE)

library(metablastr)

library(orthologr)



# system("blastp -help")
# Sys.getenv("PATH")
Sys.which("blastp")
system("blastp -version")


```

<https://drostlab.github.io/orthologr/reference/index.html>

<https://drostlab.github.io/orthologr/articles/blast.html#the-blast_rec-function>

<https://drostlab.github.io/orthologr/reference/blast_rec.html>

<https://github.com/drostlab/metablastr>

<https://rdrr.io/github/drostlab/orthologr/man/blast.html>





# Input fasta

To read fasta for some ```orthologr``` function do not use functions from ```seqinr```, rather from ```orthologr```, because they create different objects

Also, ```orthologr::blast_rec``` demands file path not variable/object

```{r,  echo=TRUE, warning=FALSE, message=FALSE}

fp = file.path('..', 'input')
list.files(fp, recursive = TRUE)

# fn1 = 'ITAG4.1_proteins.fasta'
# fn1 = 'test-ITAG4.1_proteins.fasta'
fn1 = 'a.fasta'
# Sly.proteome = seqinr::read.fasta(file = file.path(fp, fn1),
#                                   seqtype = "AA", 
#                                   as.string = FALSE, 
#                                   seqonly = FALSE)
# Sly.proteome = read.proteome(file = file.path(fp, fn1),
#                              format = 'fasta')

# PLAZA
# fn2 = 'proteome.selected_transcript.ath.fasta'
# fn2 = 'test-proteome.selected_transcript.ath.fasta'
fn2 = 'b.fasta'
# ARAPORT
# fn2 = 'Araport11_pep_20220914_representative_gene_model'
# Ath.proteome = seqinr::read.fasta(file = file.path(fp, 'PLAZA', fn2),
#                                   seqtype = "AA", 
#                                   as.string = FALSE, 
# #                                   seqonly = FALSE)
# Ath.proteome = read.proteome(file = file.path(fp, 'PLAZA', fn2),
#                              format = 'fasta')

```
 
# RBH

<https://github.com/drostlab/orthologr/issues/26>

<https://rdrr.io/github/mhahsler/rBLAST/man/BLAST.html>

<https://rdrr.io/github/drostlab/orthologr/man/blast.html>

<https://github.com/tidyverse/readr/issues/1419>



```{r blast_rec, echo=TRUE, warning=FALSE, message=FALSE}

# where the script is
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
mydir =  getwd()

# will not work, BLAST forces temp directories
# fmtp = file.path('..', 'other')
# normalizePath(fmtp)
# tools::file_path_as_absolute(fmtp)
# apparently tempdir only within working one, so this also doesnt work
# write(paste0("TMP = ",normalizePath(fmtp)), file=file.path(Sys.getenv('R_USER'), '.Renviron'))
unlink(tempdir(), recursive = TRUE, force = TRUE) # will delete directory called 'mydir'
dir.create(tempdir())


# orthologr::blast_rec(query_file    = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'), 
#            subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#            comp_cores   = 2)

Q = normalizePath(file.path(fp, 'PLAZA', fn2))
S = normalizePath(file.path(fp, fn1))

  setwd(mydir)
  
  query_file = Q
  subject_file = S
  seq_type = "protein"
  format = "fasta"
  blast_algorithm = "blastp"
  eval = "1E-10"
  max.target.seqs = 200
  path = NULL
  comp_cores = 4
  blast_params = NULL
  clean_folders = TRUE
  save.output = tempdir()
  
  file.copy(Q, paste0(tempdir()))
  file.copy(S, paste0(tempdir()))
  list.files(tempdir(), pattern = '.fasta')
  setwd(tempdir())
  

myRBH = orthologr::blast_rec(query_file = fn1
                             ,
                             subject_file = fn2
                             ,
                             seq_type = "protein"
                             ,
                             format = "fasta"
                             ,
                             blast_algorithm = "blastp"
                             ,
                             # delete_corrupt_cds = TRUE,
                             eval = "1E-10"
                             ,  
                             max.target.seqs = 200
                             ,
                             # a character string specifying the path to the BLAST program (in case you don't use the default path)
                             path = NULL
                             , 
                             comp_cores = 4
                             ,
                             # blast_params = NULL, # see whats this
                             clean_folders = TRUE
                             ,
                             # a path to the location were the BLAST output shall be stored. E.g. save.output = getwd() to store it in the current working directory, or save.output = file.path(put,your,path,here).
                             save.output =  tempdir()
                             )



```


# Write results

```{r res, echo=TRUE, warning=FALSE, message=FALSE}

setwd(dirname(rstudioapi::getActiveDocumentContext()$path))
fpo = file.path('..', 'output')
fno = 'RBHresults.txt'
data.table::fwrite(myRBH, 
                   file = file.path(fpo, fno),
                   sep = '\t')
fno = 'RBHresults.xlsx'
openxlsx::write.xlsx(myRBH, file = file.path(fpo, fno), 
                     asTable = TRUE, 
                     overwrite = TRUE)


```


# Session Info

```{r}

sessionInfo()


```



