---
title: "01_ortho"
author: "Maria, zagor"
date: "`r Sys.Date()`"
output:
  html_document:
    fig_caption: yes
    self_contained: yes
    fig_width: 12
    fig_height: 9
    toc: yes
    toc_depth: 2
    number_sections: yes
    theme: flatly
    highlight: tango
  pdf_document:
    toc: yes
    toc_depth: '2'
  word_document:
    toc: yes
    toc_depth: '2'
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(#dev = c('pdf', 'png'),  # this embeds pdf and crates scrolable blocks
                      dev = c('png'), 
                      fig.align = 'center', 
                      fig.height = 9, 
                      fig.width = 12 ,
                      warning = FALSE, message = FALSE
                      )
# options(knitr.table.format = "html")

```


1. Download ```BLAST``` for Windows (```.exe``` file) from <https://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/> and install locally

2. Add path to to NCBI blast /bin (could be ```C:\Program Files\NCBI\blast-2.14.0+\bin```) in System Enviroment Variables, e.g. on how to here <https://www.hanss.info/sebastian/post/rtools-path/>

3.Install dependencies ```iocManager::install(c("Biostrings", "GenomicFeatures", "GenomicRanges", "Rsamtools", "IRanges", "rtracklayer", "biomaRt"))```

3. Install ```metablastr``` and ```orthologr``` packages in R

```
devtools::install_github("HajkD/metablastr")

devtools::install_github("HajkD/orthologr")
```


4. Run ```system("blastp -help")``` in R Console to see if it is working



# Libraries

```{r,  echo=FALSE, warning=FALSE, message=FALSE}

rm(list = ls(all = TRUE))
gc()

set.seed(123456)


`%nin%` = Negate(`%in%`)

library(RColorBrewer)
library(magrittr)


# Install package dependencies
# BiocManager::install(c("Biostrings", "GenomicRanges", "GenomicFeatures", "Rsamtools", "rtracklayer"))

# devtools::install_github("HajkD/metablastr")
# devtools::install_github("HajkD/orthologr")

## https://heronoh.github.io/BLASTr
# remotes::install_github("heronoh/BLASTr")
# library(BLASTr)
# devtools::install_github("mhahsler/rBLAST")
# library(rBLAST)

# BiocManager::install(c("Biostrings", "GenomicFeatures", "GenomicRanges", "Rsamtools", "IRanges", "rtracklayer", "biomaRt"))

# install.packages("devtools")
# install the current version of metablastr on your system
# devtools::install_github("HajkD/metablastr", build_vignettes = TRUE, dependencies = TRUE)

library(metablastr)

library(orthologr)

system("blastp -help")

```

<https://drostlab.github.io/orthologr/reference/index.html>

<https://drostlab.github.io/orthologr/articles/blast.html#the-blast_rec-function>

<https://drostlab.github.io/orthologr/reference/blast_rec.html>

<https://github.com/drostlab/metablastr>

<https://rdrr.io/github/drostlab/orthologr/man/blast.html>




```{r}

function (query_file, subject_file, seq_type = "cds", format = "fasta", 
    blast_algorithm = "blastp", delete_corrupt_cds = TRUE, eval = "1E-5", 
    max.target.seqs = 10000, path = NULL, comp_cores = 1, blast_params = NULL, 
    clean_folders = FALSE, save.output = NULL) 
{
    orthoA <- blast_best(query_file = query_file, subject_file = subject_file, 
        eval = eval, max.target.seqs = max.target.seqs, format = format, 
        seq_type = seq_type, blast_algorithm = blast_algorithm, 
        delete_corrupt_cds = delete_corrupt_cds, path = path, 
        comp_cores = comp_cores, blast_params = blast_params, 
        save.output = save.output)
    orthoB <- blast_best(query_file = subject_file, subject_file = query_file, 
        seq_type = seq_type, eval = eval, max.target.seqs = max.target.seqs, 
        format = format, blast_algorithm = blast_algorithm, delete_corrupt_cds = delete_corrupt_cds, 
        path = path, comp_cores = comp_cores, blast_params = blast_params, 
        clean_folders = clean_folders, save.output = save.output)
     colnames(orthoB)[1:2] <- c("subject_id", "query_id")
    tryCatch({
        return(dplyr::semi_join(orthoA, orthoB, by = c("query_id", 
            "subject_id")))
    }, error = function(e) {
        stop("The BLAST tables resulting from ", query_file, 
            " and ", subject_file, " could not be joined properly to select only the reciprocal best hits.")
    })
}


```



# Input fasta

To read fasta for some ```orthologr``` function do not use functions from ```seqinr```, rather from ```orthologr```, because they create different objects

Also, ```orthologr::blast_rec``` demands file pathm not variable/object

```{r}

fp = file.path('..', 'input')
list.files(fp, recursive = TRUE)

fn1 = 'ITAG4.1_proteins.fasta'
# Sly.proteome = seqinr::read.fasta(file = file.path(fp, fn1),
#                                   seqtype = "AA", 
#                                   as.string = FALSE, 
#                                   seqonly = FALSE)
# Sly.proteome = read.proteome(file = file.path(fp, fn1),
#                              format = 'fasta')

# PLAZA
fn2 = 'proteome.selected_transcript.ath.fasta'
# ARAPORT
# fn2 = 'Araport11_pep_20220914_representative_gene_model'
# Ath.proteome = seqinr::read.fasta(file = file.path(fp, 'PLAZA', fn2),
#                                   seqtype = "AA", 
#                                   as.string = FALSE, 
# #                                   seqonly = FALSE)
# Ath.proteome = read.proteome(file = file.path(fp, 'PLAZA', fn2),
#                              format = 'fasta')

```
 
# RBH

```{r}

fmtp = file.path('..', 'other')

myRBH = orthologr::blast_rec(query_file = file.path(fp, 'PLAZA', fn2)
                             ,
                             subject_file = file.path(fp, fn1)
                             ,
                             seq_type = "protein"
                             ,
                             format = "fasta"
                             ,
                             blast_algorithm = "blastp"
                             ,
                             # delete_corrupt_cds = TRUE,
                             eval = "1E-100"
                             ,  
                             max.target.seqs = 100
                             ,
                             # a character string specifying the path to the BLAST program (in case you don't use the default path)
                             # path = file.path('C:', 'Program Files', 'NCBI', 'blast-2.14.0+', 'bin')
                             , 
                             comp_cores = 4
                             ,
                             # blast_params = NULL, # see whats this
                             clean_folders = FALSE
                             ,
                             save.output = fmtp
                             )


```


# Session Info

```{r}

sessionInfo()


```



